### PR TITLE
Fix CORS

### LINKED ISSUE #103: Implement CORS handling
Currently we don't support CORS handling (https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS), which affects JavaSript RPs. 

The problem is manifested when JS RP tries to do a cross-origin request, and it boils down to " ... the specification mandates that browsers "preflight" the request, soliciting supported methods from the server with the HTTP OPTIONS request method, and then, upon "approval" from the server, sending the actual request. Servers can also inform clients whether "credentials" (such as Cookies and HTTP Authentication) should be sent with requests. CORS failures result in errors, but for security reasons, specifics about the error are not available to JavaScript. All the code knows is that an error occurred. The only way to determine what specifically went wrong is to look at the browser's console for details..."

Problem noticed with JS OIDC client: https://www.npmjs.com/package/oidc-client

It's a bit tricky to test this since both OP and RP have to have valid TLS certs implemented (cross origin requests without proper HTTPS are limited by browsers). 

However, if one has RP with HTTPS, the problem can be quickly demonstrated using 'fetch' from browsers web console. For example, using fetch try to GET userinfo_enpoint with Authorization Bearer token header. This will trigger a preflight HTTP request, which will (currenlty) result in HTTP 500 on OP.

```
fetch('https://op.org/userinfo-endpoint', {
    method: 'GET', // *GET, POST, PUT, DELETE, etc.
    mode: 'cors', // no-cors, *cors, same-origin
    //credentials: 'include', // include, *same-origin, omit
    redirect: 'follow', // manual, *follow, error
    headers: {
       'Authorization': 'Bearer 123'
    },
  }).then(res => {console.log(res)}).catch(err => {console.log(err)});
```

image

A quick solution for HTTP OPTIONS request might be to return HTTP Header Access-Control-Allow-Origin: *, for example:
```
if ($request->getMethod() === 'OPTIONS') {
    // in real app return new Response() with proper headers
    header('Access-Control-Allow-Origin: *');
    header('Access-Control-Allow-Headers: *');
    header('Access-Control-Allow-Methods: POST, GET, OPTIONS');
    http_response_code(200);
    exit();
}
``` 

Another solution would be to enable clients to have authorized JS origins for every client, and then specify authorized origins in reponse. 

```
Access-Control-Allow-Origin: https://op.org
```

This is the way Google does it:
image

### PR DESCRIPTION
**Pull request by cicnavi**
Add new DB table and repository to store allowed origins for particular clients.

For HTTP preflight requests (HTTP OPTIONS method) to 'userinfo' endpoint, set appropriate CORS headers if particular origin is allowed.

Fixes #103

### CODE REVIEW

File: lib/Form/ClientForm.php [hunk 1]
```diff
@@ -25,6 +25,7 @@ class ClientForm extends Form
      * RFC3986. AppendixB. Parsing a URI Reference with a Regular Expression.
      */
     public const REGEX_URI = '/^[^:]+:\/\/?[^\s\/$.?#].[^\s]*$/';
+    public const REGEX_ALLOWED_ORIGIN_URL = '/http(s?)\:\/\/[^\s\/$.?#]+\.[^\s\/$.?#]+(\.[^\s\/$.?#]+)*$/i';
```
Comment (pradtke) [inline lib/Form/ClientForm.php:28]: URL regex are more readable if you use a delimiter other than '/' to indicate it is a regex. Then you need to do less escaping. e.g.  |https?\://|      maybe add a comment what this is checking. I get lost after the protocol part  :)

File: lib/Controller/OpenIdConnectUserInfoController.php [hunk 2]
```diff
@@ -99,4 +116,33 @@ private function getUser(AccessTokenEntity $accessToken)
...
+     */
+    protected function handleCors(ServerRequest $request): Response
+    {
+        $origin = $request->getHeaderLine('Origin');
+
+        if (empty($origin)) {
+            throw OidcServerException::requestNotSupported('CORS error: no Origin header present');
+        }
+
+        if (! $this->allowedOriginRepository->has($origin)) {
```
Comment (pradtke) [inline lib/Controller/OpenIdConnectUserInfoController.php:135]: so do clients not send their authorization token as a header when they send the OPTIONS call?  I just want to make sure I understand :)

File: lib/Form/ClientForm.php (hunk 1 above)
Comment (cicnavi) [reply lib/Form/ClientForm.php]: Updated the regex a bit and added a short explanation.

File: lib/Controller/OpenIdConnectUserInfoController.php (hunk 2 above)
Comment (cicnavi) [reply lib/Controller/OpenIdConnectUserInfoController.php:135]: Erm..., actually client tries to send the request using authorization header (javascript client running in the browser), but the browser stops it and the browser uses HTTP options request (preflight request) to check if the client is actually allowed to send such request... if the server responds positively (using special CORS headers), the request from the client is allowed and is sent to the server...

**Pull request merged**